File: Solutions/C++/Q9.cpp
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;
string organizingContainers(int size, vector<vector<int>> matrix) {
	string str = "";
	if (size == 2)
		if (matrix[0][0] == matrix[1][1] || matrix[0][1] == matrix[1][0])
			return "Possible";

	vector<int> columns(size + 1);
	
	for (int x=0; x<size; x++)
		for (int y=0; y<size; y++)
			if (x != y)
				columns[x] += matrix[y][x];

	sort(columns.begin(), columns.end());
	columns.erase( unique( columns.begin(), columns.end() ), columns.end() );

	for (int i=0; i<columns.size()-1; i++) {
		int head = columns[i];
		if (head > 0)
			for (int j=i+1; j<columns.size(); j++)
				columns[i] -= head;
	}

	return columns[columns.size()-1] == 0 ? "Possible" : "Impossible";
}

int main(){
	int q;
	cin >> q;
	string answer = "";
	for(int a0 = 0; a0 < q; a0++) {
		int n;
		cin >> n;
		vector<vector<int>> M(n,vector<int>(n));
		for(int M_i = 0;M_i < n;M_i++)
			for(int M_j = 0;M_j < n;M_j++)
				cin >> M[M_i][M_j];

		string result=organizingContainers(n, M);
		if (answer == "")
			answer = result;
		else
			answer = answer + "," + result;
	}
	// Do not remove below line
	cout << answer << endl;
	// Do not print anything after this line

	return 0;
}

File: Solutions/TypeScript/Q9.ts
type Input = string[];

const Q9Func = (input: Input, logging: boolean = false): string => {
	const numberOfUnsortedProblems = parseInt(input[0]);
	const unsortedProblems = input[1];

	let results: ("Possible"|"Impossible")[] = [];

	let line = 1;

	for (let i=0; i<numberOfUnsortedProblems; i++) {
		const matrixSize = parseInt(input[line]);
		line++;

		let matrix: number[][] = [];
		for (let mln=0; mln<matrixSize; mln++) {
			matrix[mln] = input[line].split(" ").map((v) => parseInt(v));
			line++;
		}

		results.push(solveMatrix(matrixSize, matrix, logging));
	}

	return results.join(",");
};

export default Q9Func;

const solveMatrix = (size: number, matrix: number[][], logging: boolean = false): "Possible"|"Impossible" => {
	if (size == 2)
		if (matrix[0][0] == matrix[1][1] || matrix[0][1] == matrix[1][0])
			return "Possible";

	let columns: number[] = new Array(size).fill(0);

	logging && console.log(matrix);

	for (let x=0; x<size; x++)
		for (let y=0; y<size; y++)
			if (x != y)
				columns[x] += matrix[y][x];

	columns = [...new Set(columns)].sort((a,b) => a-b);
	logging && console.log(columns);

	while (columns.length > 1) {
		let head = columns.shift();
		if (head && head > 0)
			for (let i=0; i<columns.length; i++)
				columns[i] -= head;
		logging && console.log(columns);
	}

	let complete = columns[0] == 0;
	return complete ? "Possible" : "Impossible";
}